drivers/periph: add basic API to read/write data to EEPROM

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR adds:
- a basic API for reading/writing data to the internal EEPROM present on certain CPUs
- support for this to the stm32l0 and stm32l1 CPU families
- a basic test application with a shell

The test application has been tested on b-l072z-lrwan1 and nucleo-l152 boards.

Since reading and writing data to EEPROM has a very similar workflow than the flashpage implementation, the base eeprom functions are implemented in the same `flashpage.c` file. This allows for example to reuse the lock/unlock functions.
The problem is that it makes this feature dependent from the flashpage feature. Any advice on a better approach will be welcome.

Maybe this PR will interest @kYc0o, @vincent-d or @haukepetersen 

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Issues/PRs references

none

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

kaspar030 commented on drivers/include/periph/eeprom.h:
```diff
@@ -0,0 +1,81 @@
+ * @param[out] data     byte address to write to
+ * @param[in]  len      the number of bytes to read
+ */
+void eeprom_read(uint32_t pos, uint8_t *data, uint8_t len);
```
Are eeproms always <256b? Otherwise, I suggest using size_t for read/write length.

kaspar030 commented:
Would you mind implementing this on a very different platform, e.g., for some atmega? That usually shows whether the interface makes sense.

aabadie commented:
> Would you mind implementing this on a very different platform, e.g., for some atmega? 

Looking at AVR Atmega1284 datasheet, this seems rather simple. I'll give it a try (already saw things that might be changed...)

aabadie commented:
@kaspar030 comments addressed + added implementation for Atmega CPUs. Tested on nucleo-l152 and Arduino Mega2560.

aabadie commented:
Do you think this one could be part of the release @kaspar030 ?

smlng commented on cpu/stm32_common/periph/flashpage.c:
```diff
@@ -201,3 +206,24 @@ void flashpage_write(int page, const void *data)
     }
 }
+
+#ifdef MODULE_PERIPH_EEPROM
```
I think this files is missing the `#include "periph/eeprom.h"`? But would also suggest to not add this to flashpage driver but rather have a separate file.

aabadie replied:
> I think this files is missing the #include "periph/eeprom.h"?

Agreed

> But would also suggest to not add this to flashpage driver but rather have a separate file.

How do you share code between different periph ? Here I'm reusing the lock/unlock functions from flashpage periph, since they are the same.

aabadie replied:
@smlng, I moved the _lock/_unlock into a shared file (between flashpage and eeprom periph drivers) and provided the eeprom peripheral driver into a separate file: now there's no dependency between flashpage and eeprom (for STMs).

smlng commented:
@aabadie are you aware of #5734, not sure if there is anything to take over?

aabadie commented:
> are you aware of #5734

Yes, but I admit I missed it when I started to work on this one.

> not sure if there is anything to take over?

I don't know. Looking at the proposed design in #5734, it seems very (too) specific so it cannot be generalized to other CPU families. I also think that it's not following RIOT's peripheral drivers designs.

aabadie commented:
The good thing with #5734 is the better error management. The API is also different but less POSIX I think.

aabadie commented:
Now that I'm thinking at it again, the interesting concept in #5734 is the use of the nvram API. I think it would be easy to adapt #8862 to nvram (following what is already done in #5734)

aabadie commented:
@kaspar030, @smlng do you have other comments here ?

smlng commented on cpu/atmega_common/periph/eeprom.c:
```diff
@@ -0,0 +1,55 @@
+
+uint8_t eeprom_read_byte(uint32_t pos)
+{
+    /* Wait for completion of previous write */
```
mhm, I thought we are reading? So maybe the comment should better state generically:
```
Wait for completion of previous operation
```

smlng commented on cpu/stm32_common/periph/eeprom.c:
```diff
@@ -0,0 +1,48 @@
+
+uint8_t eeprom_read_byte(uint32_t pos)
+{
+    DEBUG("Reading data from EEPROM at pos %lu\n", pos);
```
maybe also add `assert(pos < EEPROM_SIZE)`?

smlng commented on cpu/stm32_common/periph/eeprom.c:
```diff
@@ -0,0 +1,48 @@
+
+void eeprom_write_byte(uint32_t pos, uint8_t data)
+{
+    DEBUG("Writing data '%c' to EEPROM at pos %lu\n", data, pos);
```
same here, adding an `assert` might be a good idea?

smlng commented on cpu/stm32_common/periph/eeprom.c:
```diff
@@ -0,0 +1,48 @@
+uint8_t eeprom_read_byte(uint32_t pos)
+{
+    DEBUG("Reading data from EEPROM at pos %lu\n", pos);
+    return *(uint8_t *)(EEPROM_START_ADDR + pos);
```
question: no locking while reading needed?

smlng commented on drivers/include/periph/eeprom.h:
```diff
@@ -0,0 +1,78 @@
+ * @param[out] data     byte array to write to
+ * @param[in]  len      the number of bytes read
+ */
+void eeprom_read(uint32_t pos, uint8_t *data, size_t len);
```
IMHO these functions should return `int` for the number bytes actually read or written, or an error otherwise, applies to write functions below, too.

smlng replied:
this would also be more in line with other read/write functions in RIOT and Linux, in general.

smlng commented on drivers/periph_common/eeprom.c:
```diff
@@ -0,0 +1,48 @@
+
+#include "periph/eeprom.h"
+
+void eeprom_read(uint32_t pos, uint8_t *data, size_t len)
```
following my comment above having a proper return value, this could be something like:

```
if ((pos < EEPROM_SIZE) && (len < EEPROM_SIZE)) {
    int rc;
    for (rc = 0; (rc < len) && (pos + rc < EEPROM_SIZE); ++rc) {
        data[rc] = eeprom_read_byte(pos + rc);
    } 
    return rc;
}
return (-1);
```

applies for write function below too

kaspar030 replied:
Not the most readable. :smirk: 

Maybe just copy "len" and return it?

kaspar030 commented:
@smlng I think the API is fine as is (very slim, not much error checking). It would be very easy to add extensive error checking on top. It is difficult to remove it when the API dictates it.

aabadie commented:
@smlng, I added the `assert` as you suggested and changed the api of `eeprom_read` and `eeprom_write` functions so that they returns the number of bytes read/written (there's no complex error checking though, it's just a copy of `len`).

Let me know if it's ok for you now

aabadie commented:
@smlng, @kaspar030, comments addressed :)

mali commented on cpu/atmega_common/periph/eeprom.c:
```diff
@@ -0,0 +1,55 @@
+ */
+
+#include <stdint.h>
+
```
include <assert.h> is missing

mali commented:
Tested with Uno/Duemilanove :+1:

aabadie commented:
@smlng @kaspar030, let me know when/if I can squash this one. Thanks

smlng commented on drivers/periph_common/eeprom.c:
```diff
@@ -0,0 +1,56 @@
+{
+    assert(pos + len < EEPROM_SIZE);
+
+    size_t ret = len;
```
though it doesn't safe binary size or anything, but IMHO still a bit more readable:

```
    for (size_t i = 0; i < len; ++i) {
        data[i] = eeprom_read_byte(pos++);
    }
    return len;
```

smlng commented on drivers/periph_common/eeprom.c:
```diff
@@ -0,0 +1,56 @@
+{
+    assert(pos + len < EEPROM_SIZE);
+
+    size_t ret = len;
```
same here:

```
    for (size_t i = 0; i < len; ++i) {
        eeprom_write_byte(pos++, data[i]);
    }
    return len;
```

smlng commented on tests/periph_eeprom/main.c:
```diff
@@ -0,0 +1,110 @@
+
+#include "periph/eeprom.h"
+
+static char buffer[42];
```
buffer size should be a marco/define

smlng commented on tests/periph_eeprom/main.c:
```diff
@@ -0,0 +1,110 @@
+    uint32_t pos = atoi(argv[1]);
+    uint8_t count = atoi(argv[2]);
+
+    if (!count) {
```
also check wether it fits the buffer size

smlng commented on tests/periph_eeprom/main.c:
```diff
@@ -0,0 +1,110 @@
+        return 1;
+    }
+
+    size_t ret = eeprom_read(pos, (uint8_t *)buffer, count);
```
unsafe buffer handling, might overflow

aabadie commented:
@smlng, fixed your new comments and tested on  b-l072z-lrwan1 board. Still works well.

smlng commented on tests/periph_eeprom/main.c:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    size_t ret = eeprom_read(pos, (uint8_t *)buffer, count);
+    buffer[count] = '\0';
```
if you want to do this you must check for `count >= BUFFER_SIZE` above.

aabadie replied:
nope, the buffer has an extra byte, see [here](https://github.com/RIOT-OS/RIOT/pull/8862/files#diff-0ec3102ff4d926d871687a7b42380deaR33)

smlng replied:
oops, oversaw that 😬

kaspar030 commented on drivers/periph_common/eeprom.c:
```diff
@@ -0,0 +1,52 @@
+{
+    assert(pos + len < EEPROM_SIZE);
+
+    for (size_t i = 0; i < len; ++i) {
```
here and elsewhere: please use ```i++```.

aabadie replied:
Done and hope this is the last one ;)

smlng replied:
@kaspar030 why, if it makes a difference I always thought `++i` is better than `i++`

kaspar030 replied:
1. consistency. In RIOT, 'for.*i++' vs 'for.*++i' is about four to one.
2. In C, it doesn't make a difference. BUT "for(...;i++)' is considered ideomatic C, and thus more readable.

aabadie replied:
Ok, I'm fine with that. Now it should be all good here, right ?

kaspar030 commented on cpu/atmega_common/periph/eeprom.c:
```diff
@@ -0,0 +1,60 @@
+    assert(pos < EEPROM_SIZE);
+
+    /* Wait for completion of previous operation */
+    while (EECR & (1 << EEPE));
```
here and below: needs to be changed to ```while() {}```.

aabadie replied:
changed

kaspar030 commented on tests/periph_eeprom/Makefile:
```diff
@@ -0,0 +1,9 @@
+BOARD ?= b-l072z-lrwan1
```
why this board? Maybe chose one that is more common?

aabadie replied:
This is my dev board actually and it's also provided by iotlab ;)

aabadie commented:
and squashed

aabadie commented:
Yeah! Thanks @smlng @kaspar030 @mali for reviewing and testing.
